### PR TITLE
Fix timeline creation and tenant deletion race

### LINKED ISSUE #6255: pageserver: timeline deletion can fail on initdb ___temp files
`Error: timelines dir not empty: Timelines directory is not empty after all timelines deletion: ["initdb.tar.zst.upload-<tenant id>.___temp"]`

This happens when deletion races with creation.  Timeline deletion code should remove this file if present.

Incidents:
* https://neondb.slack.com/archives/C033RQ5SPDH/p1704217340227809
* https://neondb.slack.com/archives/C03F5SM1N02/p1704053888221219

Related:
* https://github.com/neondatabase/neon/pull/5914

### PR DESCRIPTION
**Pull request by arpad-m**
Fixes the race condition between timeline creation and tenant deletion outlined in #6255.

Related: #5914, which is a similar race condition about the uninit marker file.

Fixes #6255

### CODE REVIEW

**Comment by arpad-m** [comment]
Hmmm this CI failure is weird:

```
2024-01-10T09:32:59.7296347Z 2024-01-10 09:30:11.558 INFO [neon_fixtures.py:1878] not allowed error: 2024-01-10T09:30:04.436890Z ERROR request{method=POST path=/v1/tenant/57745c779d4051e027451cb0de7ddb10/timeline request_id=46821485-87f2-4773-8977-1051701f324b}:timeline_create{tenant_id=57745c779d4051e027451cb0de7ddb10 shard=0000 timeline_id=1e10677fe8b277809b20781e4f8e2415 lsn=None pg_version=Some(15)}: Failed to remove temporary initdb directory '/tmp/test_output/test_tenant_delete_races_timeline_creation[debug-pg15]/repo/pageserver_1/tenants/57745c779d4051e027451cb0de7ddb10/timelines/basebackup-1e10677fe8b277809b20781e4f8e2415.___temp': Directory not empty (os error 39)
2024-01-10T09:32:59.7299805Z 2024-01-10 09:30:11.559 INFO [neon_fixtures.py:1878] not allowed error: 2024-01-10T09:30:06.535732Z ERROR delete_tenant{tenant_id=57745c779d4051e027451cb0de7ddb10 shard_id=0000}: Error: timelines dir not empty: Timelines directory is not empty after all timelines deletion: ["basebackup-1e10677fe8b277809b20781e4f8e2415.___temp"]
2024-01-10T09:32:59.7302296Z 2024-01-10 09:30:11.559 INFO [neon_fixtures.py:1878] not allowed error: 2024-01-10T09:30:06.535778Z  WARN delete_tenant{tenant_id=57745c779d4051e027451cb0de7ddb10 shard_id=0000}: Marking Stopping tenant as Broken state, reason: timelines dir not empty: Timelines directory is not empty after all timelines deletion: ["basebackup-1e10677fe8b277809b20781e4f8e2415.___temp"]
2024-01-10T09:32:59.7304913Z 2024-01-10 09:30:11.559 INFO [neon_fixtures.py:1878] not allowed error: 2024-01-10T09:30:07.254569Z ERROR delete_tenant{tenant_id=57745c779d4051e027451cb0de7ddb10 shard_id=0000}: Error: timelines dir not empty: Timelines directory is not empty after all timelines deletion: ["basebackup-1e10677fe8b277809b20781e4f8e2415.___temp"]
2024-01-10T09:32:59.7306830Z 2024-01-10 09:30:11.559 INFO [neon_fixtures.py:1878] not allowed error: 2024-01-10T09:30:07.254647Z  WARN delete_tenant{tenant_id=57745c779d4051e027451cb0de7ddb10 shard_id=0000}: Tenant is already in Broken state
```

The weird part is it hitting the "Directory not empty" error [in the `remove_dir_all` call](https://github.com/neondatabase/neon/blob/979130e45e8fe611332dfed58b01d5e2a0e2c8ec/pageserver/src/tenant.rs#L3203). In theory, `remove_dir_all` allows non-empty directories.

File: test_runner/regress/test_tenant_delete.py [hunk 1]
```diff
@@ -553,3 +554,111 @@ def hit_run_failpoint():
...
+            pass
+
+    Thread(target=timeline_create).start()
+
+    def hit_initdb_upload_failpoint():
+        assert env.pageserver.log_contains(f"at failpoint {BEFORE_INITDB_UPLOAD_FAILPOINT}")
+
+    wait_until(100, 0.1, hit_initdb_upload_failpoint)
+
+    def creation_connection_timet_out():
```
Comment (problame) [inline test_runner/regress/test_tenant_delete.py:612]: typo in function name (`timet`)

**Review by problame — APPROVED** [review]

**Review by koivunej — APPROVED** [review]
I don't think adding of sync deletion code is great, but the surrounding code uses it already, alternative would had been to move away from `?` or delete the file unconditionally on the upload path. Regardless, this fixes the one more known problem so good.

**Comment by arpad-m** [comment]
two failures:
```
2024-01-12T17:14:05.5959066Z 2024-01-12 17:11:27.348 INFO [neon_fixtures.py:1889] not allowed error: 2024-01-12T17:11:20.094576Z ERROR request{method=POST path=/v1/tenant/9609dcb515f21d95bfe63670b523ff12/timeline request_id=60ae399d-11ad-4ada-8b4e-98b3c1b35635}:timeline_create{tenant_id=9609dcb515f21d95bfe63670b523ff12 shard=0000 timeline_id=0290c90f484087f0750432b8457958b3 lsn=None pg_version=Some(15)}: Failed to remove temporary initdb directory '/tmp/test_output/test_tenant_delete_races_timeline_creation[debug-pg15]/repo/pageserver_1/tenants/9609dcb515f21d95bfe63670b523ff12/timelines/basebackup-0290c90f484087f0750432b8457958b3.___temp': No such file or directory (os error 2)
2024-01-12T17:14:05.5961548Z 2024-01-12 17:11:27.348 INFO [neon_fixtures.py:1889] not allowed error: 2024-01-12T17:11:21.840806Z ERROR delete_tenant{tenant_id=9609dcb515f21d95bfe63670b523ff12 shard_id=0000}: Error: timelines dir not empty: Timelines directory is not empty after all timelines deletion: ["basebackup-0290c90f484087f0750432b8457958b3.___temp"]
2024-01-12T17:14:05.5963282Z 2024-01-12 17:11:27.348 INFO [neon_fixtures.py:1889] not allowed error: 2024-01-12T17:11:21.840857Z  WARN delete_tenant{tenant_id=9609dcb515f21d95bfe63670b523ff12 shard_id=0000}: Marking Stopping tenant as Broken state, reason: timelines dir not empty: Timelines directory is not empty after all timelines deletion: ["basebackup-0290c90f484087f0750432b8457958b3.___temp"]
2024-01-12T17:14:05.5964945Z 2024-01-12 17:11:27.348 INFO [neon_fixtures.py:1889] not allowed error: 2024-01-12T17:11:22.867241Z ERROR delete_tenant{tenant_id=9609dcb515f21d95bfe63670b523ff12 shard_id=0000}: Error: timelines dir not empty: Timelines directory is not empty after all timelines deletion: ["basebackup-0290c90f484087f0750432b8457958b3.___temp"]
2024-01-12T17:14:05.5966305Z 2024-01-12 17:11:27.348 INFO [neon_fixtures.py:1889] not allowed error: 2024-01-12T17:11:22.867307Z  WARN delete_tenant{tenant_id=9609dcb515f21d95bfe63670b523ff12 shard_id=0000}: Tenant is already in Broken state
```
and the older error:
```
2024-01-12T17:15:46.9821137Z 2024-01-12 17:12:44.217 INFO [neon_fixtures.py:1889] not allowed error: 2024-01-12T17:12:37.003733Z ERROR request{method=POST path=/v1/tenant/171b7042587207216de8cb98297527b9/timeline request_id=b503784c-8a69-4ce8-a960-f1dd093255e5}:timeline_create{tenant_id=171b7042587207216de8cb98297527b9 shard=0000 timeline_id=d6a3e57614ab642beb2d2e36b3036d29 lsn=None pg_version=Some(14)}: Failed to remove temporary initdb directory '/tmp/test_output/test_tenant_delete_races_timeline_creation[debug-pg14]/repo/pageserver_1/tenants/171b7042587207216de8cb98297527b9/timelines/basebackup-d6a3e57614ab642beb2d2e36b3036d29.___temp': Directory not empty (os error 39)
2024-01-12T17:15:46.9824334Z 2024-01-12 17:12:44.217 INFO [neon_fixtures.py:1889] not allowed error: 2024-01-12T17:12:39.241722Z ERROR delete_tenant{tenant_id=171b7042587207216de8cb98297527b9 shard_id=0000}: Error: timelines dir not empty: Timelines directory is not empty after all timelines deletion: ["basebackup-d6a3e57614ab642beb2d2e36b3036d29.___temp"]
2024-01-12T17:15:46.9826624Z 2024-01-12 17:12:44.217 INFO [neon_fixtures.py:1889] not allowed error: 2024-01-12T17:12:39.241775Z  WARN delete_tenant{tenant_id=171b7042587207216de8cb98297527b9 shard_id=0000}: Marking Stopping tenant as Broken state, reason: timelines dir not empty: Timelines directory is not empty after all timelines deletion: ["basebackup-d6a3e57614ab642beb2d2e36b3036d29.___temp"]
2024-01-12T17:15:46.9828966Z 2024-01-12 17:12:44.217 INFO [neon_fixtures.py:1889] not allowed error: 2024-01-12T17:12:39.824632Z ERROR delete_tenant{tenant_id=171b7042587207216de8cb98297527b9 shard_id=0000}: Error: timelines dir not empty: Timelines directory is not empty after all timelines deletion: ["basebackup-d6a3e57614ab642beb2d2e36b3036d29.___temp"]
2024-01-12T17:15:46.9830710Z 2024-01-12 17:12:44.218 INFO [neon_fixtures.py:1889] not allowed error: 2024-01-12T17:12:39.824697Z  WARN delete_tenant{tenant_id=171b7042587207216de8cb98297527b9 shard_id=0000}: Tenant is already in Broken state
```

File: test_runner/regress/test_tenant_delete.py [hunk 2]
```diff
@@ -553,3 +554,111 @@ def hit_run_failpoint():
...
+    BEFORE_INITDB_UPLOAD_FAILPOINT = "before-initdb-upload"
+    DELETE_BEFORE_CLEANUP_FAILPOINT = "tenant-delete-before-cleanup-remaining-fs-traces-pausable"
+
+    # Wait just before the initdb upload
+    ps_http.configure_failpoints((BEFORE_INITDB_UPLOAD_FAILPOINT, "pause"))
+
+    def timeline_create():
+        try:
+            ps_http.timeline_create(env.pg_version, tenant_id, TimelineId.generate(), timeout=1)
+            AssertionError()
```
Comment (koivunej) [inline test_runner/regress/test_tenant_delete.py:601]: unsure if this does anything, wouldn't it need `raise`? With some `assert False`, it was found that python can optimize out such statements, and the fix suggestion was to `raise RuntimeError("....")` which may or may not be more applicable.

File: test_runner/regress/test_tenant_delete.py [hunk 3]
```diff
@@ -553,3 +554,111 @@ def hit_run_failpoint():
...
+    ps_http.configure_failpoints((DELETE_BEFORE_CLEANUP_FAILPOINT, "pause"))
+
+    def tenant_delete():
+        ps_http.tenant_delete(tenant_id)
+
+    Thread(target=tenant_delete).start()
+
+    def deletion_arrived():
+        assert env.pageserver.log_contains(
+            f"cfg failpoint: {DELETE_BEFORE_CLEANUP_FAILPOINT} pause"
```
Comment (koivunej) [inline test_runner/regress/test_tenant_delete.py:630]: This is the configuration message, not "code has hit the failpoint" message, right?

Comment (arpad-m) [reply test_runner/regress/test_tenant_delete.py:630]: this is the "code has hit failpoint and is waiting" message. "code has hit failpoint" messages are still emitted in some instances.

File: test_runner/regress/test_tenant_delete.py (hunk 2 above)
Comment (koivunej) [reply test_runner/regress/test_tenant_delete.py]: raise was added..

**Comment by arpad-m** [comment]
Filed #6352 for the flaky tests. It should help with the "No such file or directory" errors, but probably won't help with the "Directory not empty" errors (that I don't have an explanation for tbh).

**Pull request merged**